### ISSUE TITLE
DataFlow BufferBlock<T> leaking

### ISSUE THREAD

**Issue by alandoherty** [dotnet/corefx#30582]
I've had to fix numerous problems in my projects using `BufferBlock<T>` leaking memory, I can't tell if this is by design or if it's actually unintended behaviour. Creating a `BufferBlock<T>` without completing it seems to leak both memory and scheduled tasks slowly grinding the application to a halt.

Quick example to trigger problem, results in memory leak:

```csharp
class Program
{
    static void Main(string[] args)
    {
        int gcStep = 0;
        
        while(true) {
            BufferBlock<Test> test = new BufferBlock<Test>();
            gcStep++;

            if (gcStep % 1000 == 0)
                GC.Collect(GC.MaxGeneration, GCCollectionMode.Forced, true);
        }
    }
}

class Test { }
```

Quick example to remedy problem:

```csharp
class Program
{
    static void Main(string[] args)
    {
        int gcStep = 0;
        
        while(true) {
            BufferBlock<Test> test = new BufferBlock<Test>();
            test.Complete();
            gcStep++;

            if (gcStep % 1000 == 0)
                GC.Collect(GC.MaxGeneration, GCCollectionMode.Forced, true);
        }
    }
}

class Test { }
```

It's not obvious in the documentation that this occurs, and seems like a really easy mistake that can have drastic effects on async performance and memory. My guess is that the Task is keeping the `BufferBlock<T>` reference alive.

The task which is leaked for every buffer block:

```
Not Flagged		68006	Scheduled	[Scheduled and waiting to run]	Task.ContinueWith: <.ctor>b__7_2
```

**Comment by Clockwork-Muse** [comment]
...what makes you think that BufferBlock is leaking anything, Task or otherwise?

**Comment by alandoherty** [comment]
I stripped the application down to just that code, I've found this to be a problem in some of my other code but it occurs even when I call BufferBlock<Test> constructor with nothing else (in .NET Console App project). I find that this problem leaks memory and tasks as shown below. It is resolved if you call `Complete` however.

![image](https://user-images.githubusercontent.com/2723639/41749778-56a2f682-75b0-11e8-8331-8b1b2efba59d.png)

![image](https://user-images.githubusercontent.com/2723639/41749854-c305ff36-75b0-11e8-8031-47cd4a3e422e.png)

**Comment by svick** [comment]
When I run your code using the `dotnet` command line on .Net Core 2.1 or .Net Framework 4.7.2 and System.Threading.Tasks.Dataflow 4.9.0 on Windows 10, I don't see the behavior you're describing: memory usage according to Task Manager stays stable at few megabytes.

The only thing I can think of: are you using a recent version of the System.Threading.Tasks.Dataflow package (and *not* the deprecated Microsoft.Tpl.Dataflow package)?

**Comment by alandoherty** [comment]
I'm using .NET Core 2.0 currently, let me try 2.1 and see if that fixes the issue. I'm using the `System.Threading.Tasks.Dataflow` package, and there isn't anything special about my setup. I've found this issue to occur on Linux as well.

**Comment by alandoherty** [comment]
Just tested and it doesn't occur with `dotnet` but it does occur when running from Visual Studio 2017 with either Debug or Release configurations.

**Comment by stephentoub** [comment]
> it does occur when running from Visual Studio 2017 and .NET Core 2.0/2.1 with both Debug or Release configurations

F5 or ctrl-F5? i.e. with the debugger attached or not?

**Comment by alandoherty** [comment]
Only with the debugger attached it seems.

**Comment by stephentoub** [comment]
When the debugger is attached, Tasks are tracked in a static table and kept alive until they're completed, e.g.
https://github.com/dotnet/coreclr/blob/7196a7f62ba4047ddd51523cc4a253ba44201033/src/System.Private.CoreLib/src/System/Threading/Tasks/TaskContinuation.cs#L304-L307

**Comment by alandoherty** [comment]
Is there any way to disable this behaviour? Or mitigate rouge libraries from creating tasks which never complete? It makes it quite difficult to debug memory/performance issues since it adds noise when you specifically are aiming to look for those sorts of issues.

**Comment by svick** [comment]
@stephentoub Would it make sense to track them using weak pointers (or a similar approach) instead? That way, abandoned tasks could be GCed.

**Comment by Clockwork-Muse** [comment]
> Or mitigate rouge libraries from creating tasks which never complete?

Um, no.  You're asking to solve the [halting problem](https://en.wikipedia.org/wiki/Halting_problem).
You could probably clean up abandoned tasks that aren't started yet (cannot do work), or abandoned and running that don't have external references or modify no outside-the-task-state (no visible effect) - the behavior of the program suggests that at least the first is being performed.

**Comment by Clockwork-Muse** [comment]
@svick - to me it sounds like the goal was to prevent them just immediately being GC'd, so you could track down where they came from, because creating a large number of abandoned tasks would normally be considered a programmer error.

@alandoherty - Is the code you're working on really creating a block in a tight loop like that?  And normally you are expected to call `Complete` (and allow it to clean up), so you should at least be doing that.

**Comment by stephentoub** [comment]
> Would it make sense to track them using weak pointers (or a similar approach) instead? That way, abandoned tasks could be GCed.

To answer that I will need to go back and review what purpose the static table serves; I don't remember at this point.

**Comment by alandoherty** [comment]
I think there's a couple of problems here, firstly it's really not obvious at all in the documentation for `BufferBlock<T>` that you must Complete the block. I got the idea that it would be safe to discard unused blocks without completing, especially since it's offered in some places as a asynchronous concurrent queue.

Also it was my understanding that it was good practice to always complete tasks eventually, so perhaps the weak reference and auto completion of the task would be an idea? And it would be nice if there was a way to turn off this static table, so that you can prevent uncompleted tasks from causing issues during debugging for other issues.

**Comment by stephentoub** [comment]
@gregg-miskelly, does VS still use the private Task.GetActiveTaskFromId method?  What would the impact be if we were to remove it?  Or, what would the impact be if VS passed in the id of a Task that would have otherwise been garbage collected by then?

**Comment by gregg-miskelly** [comment]
CC @r-ramesh

From reading the code what would happen is that VS will have a row in the tasks window that has whatever original name the task had (from the ETW event) and no location. We might want a new ETW event to notify that a uncompleted task was GC'ed?

To answer your question from Friday - no, we don't have a knob to prevent setting s_asyncDebuggingEnabled.

**Comment by stephentoub** [comment]
> We might want a new ETW event to notify that a uncompleted task was GC'ed?

I don't think we can do that. This would be *super* expensive, as Task would need a finalizer, and even if we GC.SuppressFinalize'd it if the ETW event wasn't enabled, there's still measurable cost there. In .NET Core 2.1 we added something similar for async methods, but there the runtime controls the instantiation of the async method boxing object, and thus if events are enabled, it can choose to construct one that has a finalizer, whereas the one normally constructed won't.
https://github.com/dotnet/coreclr/blob/df78ae72d5cf3f2c2dbe4ff972732d418497f9bc/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs#L505-L514

**Comment by gregg-miskelly** [comment]
Perhaps we could solve it by just checking for a distinctive exception when we called `Task.GetActiveTaskFromId` to understand that the task has been collected...

**Comment by stephentoub** [comment]
> Perhaps we could solve it by better handling a null result from Task.GetActiveTaskFromId to understand that the task has been collected

If that's feasible, then we could change it to use a weak collection.

Is it?
